### DISCUSSION TITLE
EG4 6000EX-48HV - I can only get one command to work, that is Q1

### DISCUSSION THREAD

**Discussion by BobbyCannon** [jblance/mpp-solar#309] [General]
I'm currently trying to get information from my EG4 6000EX-48HV. The only command I can get work is "Q1". No other commands will work. I've tried QID, QMOD, QPI, QPIGS, etc. All those command returns back "(NAKss".

I've tried sending commands with and without a CRC but only Q1 seems to work.

Ex of what I'm sending for
* Works: Q1 is 0x51, 0x31, 0x1B, 0xFC, 0x0D
* Fails: QPIGS is 0x51, 0x50, 0x49, 0x47, 0x53, 0xB7, 0xA9, 0x0D

Just wondering if anyone has successfully queried a 6000EX yet.

**Comment by Saentist** [comment]
What protocol you use?

**Reply by BobbyCannon** [reply]
I'm actually not using the app at all. Which your question made me think that I should just setup an environment and try the app itself. I don't have a linux box nor do I have any available hardware though.

I may try WSL (Windows Subsystem for Linux) and see if I could get it running from there.

But I will spend more time in the protocol files. I have no idea what protocol should be used. I'm talking to my inverter directly using nothing more than a C# SerialPort. Raw as can be.

I'll take the time and see if I could set it up.

What I'm trying to do is very similar to the following post. https://github.com/jblance/mpp-solar/discussions/284.

**Reply by Saentist** [reply]
What version of mpp-solar you use?
`mpp-solar -v`

**Reply by BobbyCannon** [reply]
@Saentist I'm not using mpp-solar at all. I'm just looking at the code base and trying to figure out the protocol / communication expectations for my EG4 6000EX-48HV. I'm thinking I need to just setup a box and play around with mpp-solar but I'm in Windows and don't really have access to more hardware to setup another box. Well maybe that's not 100% true but I really don't want to setup another box. LOL

I do have an old mac mini laying around... Hum, maybe I should install Ubuntu on it... wonder how hard that would be?

**Reply by Saentist** [reply]
Win 10 support WSL so no problem to use it.
Try is multisibcontrol can handle it.

**Reply by BobbyCannon** [reply]
Oh man, got WSL setup, got MPP-SOLAR installed and boom... found out WSL2 doesn't support USB with much work. LOL

**Reply by BobbyCannon** [reply]
I manually tested every command in every protocol. Here is what I found.

Q1, QVFW3, QOPPT, QCHPT, QMN, QBMS all work, at least return data.

The following commands all return (NAKss.

"PQSE", "QALL", "QBEQI", "QDI", "QED[yyyy][mm][dd]", "QEM", 
"QEM[yyyy][mm]", "QET", "QEY", "QEY[yyyy]", "QFLAG", "QGMN",
"QLED", "QLT", "QMCHGCR", "QMOD", "QMUCHGCR", "QOPM", "QPGS",
"QPI", "QPIGS", "QPIGS2", "QPIRI", "QPIWS", "QSID", "QT", 
"SOC", "VERFW", "QLY2022", "QLM202201", "QLD20220205", "QBOOT",
"QVFW", "QVFW2", "QVFW4", "PI", "ID", "VFW", "ET", "PIRI", "GS",
"QPGS0", "QPGS1"

QMN returns "VMIII-6000" so I'm off to see if I can google up more data.

**Reply by BobbyCannon** [reply]
OK, figured it out. It seems the protocol to use with EG4 6000EX-48HV is pi18. Using those commands are returning all kinds of data. Woot!

**Reply by Saentist** [reply]
If you get latest firmware file, there you can see all commands supported.
https://eg4electronics.com/wp-content/uploads/2022/12/New-6k-Firmware.zip
there is a file **Inv.out** with utf-8 encoding can see commands included
![image](https://user-images.githubusercontent.com/16975883/217236291-523541db-a631-4806-81d4-92408dc7ad05.png)

strange commands I see
Q2
Q3
Q7
QGPRS
QOPC
QBMS
QABC
QSCIR

**Reply by BobbyCannon** [reply]
Most excellent. Yeah, I have no problems now. I'm just interested in building out my custom "Farm" app for managing my farm. This is exactly what I was looking for. I appreciate all the help!

**Reply by Saentist** [reply]
Can you post `raw` outputs of this **`Query`** commands?
Some of them are really new.
